src/mallocTracer.cpp

```diff
@@ -0,0 +1,230 @@
+static realloc_t _orig_realloc = NULL;
+
+typedef void (*free_t)(void*);
+static free_t _orig_free = NULL;
+
+static void* malloc_hook(size_t size) {
+    void* ret = _orig_calloc(1, size);
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* malloc(size_t size) {
+    if (unlikely(!_orig_malloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return malloc_hook(size);
+    }
+    return _orig_malloc(size);
+}
+#endif // __linux__
+
+static void* calloc_hook(size_t num, size_t size) {
+    void* ret = _orig_calloc(num, size);
+    if (likely(ret && num && size)) {
+        MallocTracer::recordMalloc(ret, num * size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* calloc(size_t num, size_t size) {
+    if (unlikely(!_orig_calloc)) {
+        // In some libc versions, dlsym may call calloc on error case in the ADDRESS_OF macro.
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return calloc_hook(num, size);
+    }
+    return _orig_calloc(num, size);
+}
+#endif // __linux__
+
+static void* realloc_hook(void* addr, size_t size) {
+    void* ret = _orig_realloc(addr, size);
+    if (likely(ret && addr)) {
+        MallocTracer::recordFree(addr, OS::nanotime());
+    }
+
+    if (likely(ret && size)) {
+        MallocTracer::recordMalloc(ret, size, OS::nanotime());
+    }
+    return ret;
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void* realloc(void* addr, size_t size) {
+    if (unlikely(!_orig_realloc)) {
+        return NULL;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return realloc_hook(addr, size);
+    }
+    return _orig_realloc(addr, size);
+}
+#endif // __linux__
+
+static void free_hook(void* addr) {
+    _orig_free(addr);
+    MallocTracer::recordFree(addr, OS::nanotime());
+}
+
+#ifdef __linux__
+extern "C" WEAK DLLEXPORT void free(void* addr) {
+    if (unlikely(!_orig_free)) {
+        return;
+    }
+
+    if (likely(MallocTracer::initialized())) {
+        return free_hook(addr);
+    }
+    return _orig_free(addr);
+}
+#endif // __linux__
+
+u64 MallocTracer::_interval;
+volatile u64 MallocTracer::_allocated_bytes;
+
+Mutex MallocTracer::_patch_lock;
+int MallocTracer::_patched_libs = 0;
+bool MallocTracer::_initialized = false;
+
+__attribute__((constructor)) static void getOrigAddresses() {
+    // Store these addresses, regardless of MallocTracer being enabled or not.
+    _orig_calloc = ADDRESS_OF(calloc);
+    _orig_free = ADDRESS_OF(free);
+    _orig_malloc = ADDRESS_OF(malloc);
+    _orig_realloc = ADDRESS_OF(realloc);
+}
+
+bool MallocTracer::initialize() {
+    if (!__sync_bool_compare_and_swap(&_initialized, false, true)) {
+        return false;
+    }
+
+    CodeCache* lib = Profiler::instance()->findLibraryByName("libasyncProfiler");
```

Comment:
It's not necessarily called like this. A library can have an arbitrary name, especially when extracted as a temporary file. Use `findLibraryByAddress` instead.